lib/spack/spack/relocate.py

```diff
@@ -97,70 +97,93 @@ def _patchelf():
+    Args:
+        path (str): full path to the executable or library
 
-    rpaths = list()
+    Return:
+        RPATHs as a list of strings.
+    """
+    # If we're relocating patchelf itself, use it
+    patchelf_path = path if path.endswith("/bin/patchelf") else _patchelf()
+    patchelf = executable.Executable(patchelf_path)
+
+    output = ''
     try:
-        output = patchelf('--print-rpath', '%s' %
-                          path_name, output=str, error=str)
-        rpaths = output.rstrip('\n').split(':')
+        output = patchelf('--print-rpath', path, output=str, error=str)
+        output = output.strip('\n')
     except executable.ProcessError as e:
-        msg = 'patchelf --print-rpath %s produced an error %s' % (path_name, e)
-        tty.warn(msg)
-    return rpaths
+        msg = 'patchelf --print-rpath {0} produced an error [{1}]'
+        tty.warn(msg.format(path, str(e)))
 
+    return output.split(':') if output else []
 
-def get_relative_elf_rpaths(path_name, orig_layout_root, orig_rpaths):
-    """
-    Replaces orig rpath with relative path from dirname(path_name) if an rpath
-    in orig_rpaths contains orig_layout_root. Prefixes $ORIGIN
-    to relative paths and returns replacement rpaths.
-    """
-    rel_rpaths = []
-    for rpath in orig_rpaths:
-        if re.match(orig_layout_root, rpath):
-            rel = os.path.relpath(rpath, start=os.path.dirname(path_name))
-            rel_rpaths.append(os.path.join('$ORIGIN', '%s' % rel))
-        else:
-            rel_rpaths.append(rpath)
-    return rel_rpaths
 
+def _make_relative(start_path, path_root, paths):
+    """Return a list where any path in ``paths`` that starts with
+    ``path_root`` is made relative to ``dirname(start_path)``.
 
-def get_normalized_elf_rpaths(orig_path_name, rel_rpaths):
-    """
-    Normalize the relative rpaths with respect to the original path name
-    of the file. If the rpath starts with $ORIGIN replace $ORIGIN with the
-    dirname of the original path name and then normalize the rpath.
-    A dictionary mapping relativized rpaths to normalized rpaths is returned.
-    """
-    norm_rpaths = list()
-    for rpath in rel_rpaths:
-        if rpath.startswith('$ORIGIN'):
-            sub = re.sub(re.escape('$ORIGIN'),
-                         os.path.dirname(orig_path_name),
-                         rpath)
-            norm = os.path.normpath(sub)
-            norm_rpaths.append(norm)
-        else:
-            norm_rpaths.append(rpath)
-    return norm_rpaths
+    After a path is made relative it is prefixed with the ``$ORIGIN``
+    string.
 
+    Args:
+        start_path (str): path from which the starting directory
+            is extracted
+        path_root (str): root of the relative paths
+        paths: paths to be examined
 
-def set_placeholder(dirname):
+    Returns:
+        List of relative paths
     """
-    return string of @'s with same length
+    start_directory = os.path.dirname(start_path)
```

Comment:
@gartung Is there any reason why we take the `dirname` here rather than asking for the first argument being a directory?

Reply:
@scheibelp I agree with what you say after having looked better at the contexts in which this function is called.